File: TopInterview150/SlidingWindow/SubstringWithConcatenationOfAllWords/solution.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        int wordLen = words[0].size();
        int numWords = words.size();
        int windowSize = wordLen*numWords;
        vector<int> output;

        //check boundary cases
        if (windowSize > s.length())
        {
            return output;
        }

        //create hashmap
        unordered_map<string, int> targetWordsMap;
        //populate hashmap with words vector and counts
        for (int i = 0; i < words.size(); i++)
        {
            if (targetWordsMap.find(words[i]) != targetWordsMap.end())
            {
                targetWordsMap[words[i]]++;
            }
            else targetWordsMap.insert({words[i], 1});
        }

        //iterate through the possible starting positions of the sliding window
        //while checking for matches/lack thereof
        for (int i = 0; i < wordLen; i++)
        {
            int left = i;
            int right = i;
            unordered_map<string, int> currCountMap;
            int count = 0;

            //slide the windoe right wordLen at a time
            /* for example words[] = foo, bar, s.length == 10, 
                it will check starting indexes:
                0, 3, 6,    
                1, 4, 7,
                2, 5
                which is really odd but apparently optimal
            */
            while (right + wordLen <= s.length())
            {
                string word = s.substr(right, wordLen);
                right += wordLen;

                //if not in map
                if (targetWordsMap.find(word) == targetWordsMap.end())
                {
                    currCountMap.clear();
                    left = right;//because right wasn't viable
                    count = 0;//reset count
                }
                else //otherwise add it to the correct count
                {
                    currCountMap[word]++;//automatically handles new word additions
                    count++;

                    while (currCountMap[word] > targetWordsMap[word])
                    {
                        string leftWord = s.substr(left, wordLen);
                        currCountMap[leftWord]--;
                        count --;
                        left += wordLen;
                    }
                    if (count == numWords)
                        output.push_back(left);
                }
            }

        }
        return output;
    }
};

File: TopInterview150/SlidingWindow/SubstringWithConcatenationOfAllWords/solution.py
class Solution(object):
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        output = []
        numWords = len(words)
        wordLen = len(words[0])
        windowLength = numWords*wordLen

        if len(s) < windowLength:
            return output
        
        targetCountsMap = {}

        for word in words:
            if word in targetCountsMap:
                targetCountsMap[word]+=1
            else:
                targetCountsMap[word] = 1

        for i in range(wordLen):
            """iterate over each possible starting index of the first word"""
            left = i
            right = i
            currCountMap = {}
            count = 0

            while (right + wordLen <= len(s)):
                word = s[right:right+wordLen]
                right += wordLen

                if word not in targetCountsMap:
                    currCountMap.clear()
                    count = 0
                    left = right
                else:
                    if word in currCountMap:
                        currCountMap[word] +=1
                    else:
                        currCountMap[word] = 1
                    count+=1
                    while currCountMap[word] > targetCountsMap[word]:
                        leftWord = s[left:left+wordLen]
                        currCountMap[leftWord]-=1
                        count-=1
                        left += wordLen
                    if count == numWords:
                        output.append(left)
        return output